src/engraving/libmscore/harppedaldiagram.cpp

```diff
@@ -0,0 +1,250 @@
+}
+
+HarpPedalDiagram::HarpPedalDiagram(const HarpPedalDiagram& h)
+    : TextBase(h)
+{
+    _pedalState = h._pedalState;
+    _isDiagram = h._isDiagram;
+}
+
+void HarpPedalDiagram::layout()
+{
+    updateDiagramText();
+
+    TextBase::layout();
+}
+
+void HarpPedalDiagram::setPedalState(std::vector<PedalPosition> state)
+{
+    _pedalState = state;
+}
+
+void HarpPedalDiagram::setIsDiagram(bool diagram)
+{
+    _isDiagram = diagram;
+}
+
+void HarpPedalDiagram::setPedal(HarpString harpString, PedalPosition pedal)
+{
+    _pedalState.at(harpString) = pedal;
+}
+
+void HarpPedalDiagram::updateDiagramText()
+{
+    String diagram = String("");
+    if (_isDiagram) {
+        for (int idx = 0; idx < _pedalState.size(); idx++) {
+            if (idx == 3) {
+                // insert separator
+                diagram.append(String("<sym>harpPedalDivider</sym>"));
+            }
+            switch (_pedalState[idx]) {
+            case PedalPosition::FLAT:
+                diagram.append(String("<sym>harpPedalRaised</sym>"));
+                break;
+            case PedalPosition::NATURAL:
+                diagram.append(String("<sym>harpPedalCentered</sym>"));
+                break;
+            case PedalPosition::SHARP:
+                diagram.append(String("<sym>harpPedalLowered</sym>"));
+                break;
+            }
+        }
+    } else {
+        // find difference between pedal state of previous diagram and this one
+        // if no previous, assume starting from all natural
+        std::vector<PedalPosition> prevState;
+        HarpPedalDiagram* prevDiagram = searchPrevDiagram();
+
+        if (prevDiagram != nullptr) {
+            prevState = prevDiagram->getPedalState();
+        } else {
+            prevState = { PedalPosition::FLAT, PedalPosition::FLAT, PedalPosition::FLAT, PedalPosition::FLAT,
+                          PedalPosition::FLAT, PedalPosition::FLAT, PedalPosition::FLAT };
+        }
+
+        for (int idx = 0; idx < _pedalState.size(); idx++) {
+            if (_pedalState[idx] != prevState[idx]) {
+                String strName = getStringName(static_cast<HarpString>(idx));
+                switch (_pedalState[idx]) {
+                case PedalPosition::FLAT:
+                    diagram.append(strName + String("<sym>accidentalFlat</sym>, "));
+                    break;
+                case PedalPosition::NATURAL:
+                    diagram.append(strName + String("<sym>accidentalNatural</sym>, "));
+                    break;
+                case PedalPosition::SHARP:
+                    diagram.append(strName + String("<sym>accidentalSharp</sym>, "));
+                    break;
+                }
+            }
+        }
+        // trailing comma
+        if (diagram.size() != 0) {
+            diagram.truncate(diagram.size() - 2);
+        }
+    }
+
+    setXmlText(diagram);
+}
+
+const String HarpPedalDiagram::getStringName(HarpString str)
```

Comment:
This method should be static and should be declared directly in this file. Also, HarpString should be renamed to HarpStringType (and the argument to "type" accordingly). After these manipulations the method itself should be renamed to harpStringTypeToString